src/operator/tensor/elemwise_binary_op-inl.h

```diff
@@ -374,6 +374,72 @@ void ElemwiseBinaryOp::CsrCsrOp(mshadow::Stream<cpu> *s,
   }
 }
 
+template<typename OP>
+struct ElemwiseDnsZeroKernel {
+  template<typename DType>
+  static void inline Map(int i, const OpReqType req, DType* out, const DType* dns_data,
+                         const nnvm::dim_t num_rows, const nnvm::dim_t num_cols) {
+    if (i < num_rows*num_cols) {
+      KERNEL_ASSIGN(out[i], req, OP::Map(dns_data[i], DType(0.0f)));
+    }
+  }
+};
+
+template<typename OP>
+struct ElemwiseDnsCsrDnsKernel {
+  template<typename DType, typename IType, typename CType>
+  static void inline Map(int i, const OpReqType req, DType* out, DType* dns_data,
+                         const DType* csr_data, const IType* csr_indices, const CType* csr_indptr,
+                         const nnvm::dim_t num_rows, const nnvm::dim_t num_cols) {
+    if (i < num_rows) {
+      for (int j = csr_indptr[i]; j < csr_indptr[i+1]; ++j) {
+        KERNEL_ASSIGN(out[i * num_cols + csr_indices[j]], req,
+                      OP::Map(dns_data[i * num_cols + csr_indices[j]], csr_data[j]));
+      }
+    }
+  }
+};
+
+/*! \brief DNS -op- CSR binary operator for non-canonical NDArray */
+template<typename DType, typename IType, typename CType, typename OP>
+void ElemwiseBinaryOp::DnsCsrDnsOp(mshadow::Stream<cpu> *s,
+                                   const nnvm::NodeAttrs &attrs,
+                                   const OpContext &ctx,
+                                   const NDArray &dns,
+                                   const NDArray &csr,
+                                   const OpReqType req,
+                                   const NDArray &output,
+                                   const bool reverse) {
+  using namespace mshadow;
+  using namespace mxnet_op;
+  CHECK_EQ(dns.storage_type(), kDefaultStorage);
+  CHECK_EQ(csr.storage_type(), kCSRStorage);
+  const nnvm::dim_t num_csr_rows = csr.shape()[0];
+  const nnvm::dim_t num_csr_cols = csr.shape()[1];
+  mxnet_op::Kernel<ElemwiseDnsZeroKernel<OP>, cpu>::Launch(
+    s, output.data().Size(), req, output.data().dptr<DType>(), dns.data().dptr<DType>(),
+    num_csr_rows, num_csr_cols);
+  TBlob csr_data = csr.data();
+  TBlob csr_indices = csr.aux_data(csr::kIdx);
+  TBlob csr_indptr = csr.aux_data(csr::kIndPtr);
+  MSHADOW_SGL_DBL_TYPE_SWITCH(csr_data.type_flag_, DataType, {
```

Comment:
The function is already templated. DType === DataType, right? 
I think the part of the template "typename DType, typename IType, typename CType" can be removed, which makes it easier for other operators to call this function. 
I noticed that RspRspOp is also templated. We probably canmove the DType switches inside the implementation to make the code cleaner, but that will be out of the scope of this PR.

Reply:
Done with this function, will discuss with you on the refactor later